packages/next/build/plugins/collect-plugins.ts

```diff
@@ -0,0 +1,239 @@
+  const pluginMetaData: {
+    name: string
+    'required-env': string[]
+  } = pluginPackageJson.nextjs
+
+  if (!pluginMetaData) {
+    exitWithError('Next.js plugins need to have a "nextjs" key in package.json')
+  }
+
+  if (!pluginMetaData.name) {
+    exitWithError(
+      'Next.js plugins need to have a "nextjs.name" key in package.json'
+    )
+  }
+
+  // TODO: add err.sh explaining requirements
+  let middleware: string[] = []
+  try {
+    middleware = await readdir(path.join(pkgDir, 'src'))
+  } catch (err) {
+    if (err.code !== 'ENOENT') {
+      console.error(err)
+    }
+    exitWithError(
+      `Failed to read src/ directory for Next.js plugin: ${pluginMetaData.name}`
+    )
+  }
+
+  // remove the extension from the middleware
+  middleware = middleware.map(item => {
+    const parts = item.split('.')
+    parts.pop()
+    return parts.join('.')
+  })
+
+  const invalidMiddleware: string[] = []
+
+  for (const item of middleware) {
+    if (!VALID_MIDDLEWARE.includes(item)) {
+      invalidMiddleware.push(item)
+    }
+  }
+
+  if (invalidMiddleware.length > 0) {
+    console.error(
+      `Next.js Plugin: ${
+        pluginMetaData.name
+      } listed invalid middleware ${invalidMiddleware.join(', ')}`
+    )
+  }
+
+  // TODO: investigate requiring plugins' env be prefixed
+  // somehow to prevent collision
+  if (!Array.isArray(pluginMetaData['required-env'])) {
+    exitWithError(
+      'Next.js plugins need to have a "nextjs.required-env" key in package.json'
+    )
+  }
+
+  const missingEnvFields: string[] = []
+
+  for (const field of pluginMetaData['required-env']) {
+    if (typeof env[field] === 'undefined') {
+      missingEnvFields.push(field)
+    }
+  }
+
+  if (missingEnvFields.length > 0) {
+    exitWithError(
+      `Next.js Plugin: ${
+        pluginMetaData.name
+      } required env ${missingEnvFields.join(
+        ', '
+      )} but was missing in your \`next.config.js\``
+    )
+  }
+
+  return {
+    middleware,
+    directory: pkgDir,
+    requiredEnv: pluginMetaData['required-env'],
+    version: pluginPackageJson.version,
+    pluginName: pluginMetaData.name,
+    pkgName: pluginPackageJson.name,
+  }
+}
+
+type SeparatedPlugins = {
+  appMiddlewarePlugins: PluginMetaData[]
+  documentMiddlewarePlugins: PluginMetaData[]
+}
+
+// clean package name so it can be used as variable
+export const getPluginId = (pkg: string): string => {
+  pkg = pkg.replace(/\W/g, '')
+
+  if (pkg.match(/^[0-9]/)) {
+    pkg = `_${pkg}`
+  }
+  return pkg
+}
+
+type PluginConfig =
+  | string
+  | {
+      name: string
+      config: { [name: string]: any }
+    }
+
+async function _collectPlugins(
+  dir: string,
+  env: ENV_OPTIONS,
+  pluginsConfig: PluginConfig[] | undefined
+): Promise<PluginMetaData[]> {
+  const hasPluginConfig = Array.isArray(pluginsConfig)
+  let nextPluginNames: string[] = []
+
+  if (hasPluginConfig) {
+    console.log('Found plugins config, auto detecting plugins disabled')
```

Comment:
Instead of this message, maybe warn on an unused plugin?

Reply:
Updated to show the auto-detected packages that were skipped